src/Neo/Ledger/Blockchain.cs

```diff
@@ -414,26 +416,47 @@ private void Persist(Block block)
                     all_application_executed.Add(application_executed);
                     transactionStates = engine.GetState<TransactionState[]>();
                 }
-                DataCache clonedSnapshot = snapshot.CreateSnapshot();
-                // Warning: Do not write into variable snapshot directly. Write into variable clonedSnapshot and commit instead.
+
+                var tasks = new List<Task<(TransactionState, DataCache, ApplicationExecuted)>>();
+
+                // Warning: Do not write into variable snapshot directly.
+                // Write into variable clonedSnapshot and commit instead.
                 foreach (TransactionState transactionState in transactionStates)
                 {
-                    Transaction tx = transactionState.Transaction;
-                    using ApplicationEngine engine = ApplicationEngine.Create(TriggerType.Application, tx, clonedSnapshot, block, system.Settings, tx.SystemFee);
-                    engine.LoadScript(tx.Script);
-                    transactionState.State = engine.Execute();
-                    if (transactionState.State == VMState.HALT)
+                    DataCache txSnapshot = snapshot.CreateSnapshot();
+                    var task = OnExecuteTransactionAsync(system, block, txSnapshot, transactionState);
+                    tasks.Add(task);
+                }
+
+                // var readSet = new HashSet<StorageKey>();
+                var writeSet = new HashSet<StorageKey>();
+
+                var randomUsed = false;
+                foreach (var task in tasks)
```

Comment:
Can you combine these two loops? I see you using `OnExecuteTransactionAsync` to do the loading. But I think if you were to batch or even just have a callback method for `OnExecuteTransactionAsync` you should be good and maybe speed up some more with one loop instead of two loops. Or you could use `Task.WhenAll` to keep sync.

Reply:
They are different, parallel execution, sequential commit.